[IR] Implemented collapsing of fold trees

joroKr21 opened the pull request:
This includes:
- Fusion of multiple independent folds (banana split)
- Squashing of intermediate monadic ops into folds

After this, fold-group-fusion can be implemented.

Resolves #232

aalexandrov commented:
The "monadic squash" fusion can be defined with a single rewrite rule if we use Emma Core instead of LNF using one rewrite rule. This will probably also work faster because chains of `map` / `flatMap` and `withFilter` would have been squashed to a single comprehension during the comprehension normalization phase.

In the following `$id` matches a subtree.

```scala
val ys = {
  x_1 <- $rhs_1
  x_2 <- $rhs_2
  // ...
  x_n <- $rhs_n
  if $p_1
  // ...
  if $p_n
} yield $ret

val zs = ys.fold(z1)(s1, u1)
```

if the dependency chain between `xi` induced by uses in `rhs_i` is linear, we can rewrite this as follows

```scala
val s2 = x_1 => {
  val us = for {
    x_2 <- $rhs_2
    // ...
    x_n <- $rhs_n
    if $p_1
    // ...
    if $p_n
  } yield $ret
  us.fold(z1)(s1, u1)
}
val ys = fold(z1)(s1, u1)
```

joroKr21 commented:
That's a good suggestion. You would still have to take care of the special case when the comprehension has only one generator though. You can't write:
```scala
val s2 = x_1 => {
  val us = for {
    if $p_1
    // ...
    if $p_n
  } yield $ret
  us.fold(z1)(s1, u1)
}
```

But you could write:
```scala
val s2 = x_1 => {
  if ($p_1 && ... && $p_n) $ret
  else z1
}
```

aalexandrov commented on emma-spark/src/main/scala/org/emmalanguage/api/SparkDataset.scala:
```diff
@@ -214,4 +214,15 @@ object SparkDataset {
+    xs: DataBag[A], key: A => K, sng: A => B, uni: (B, B) => B
+  )(implicit spark: SparkSession): DataBag[(K, B)] = xs match {
+    case xs: SparkRDD[A] => spark.createDataset(xs.rep
+      .map(x => key(x) -> sng(x)).reduceByKey(uni))
```
Call `SparkRDD.foldGroup` in order to avoid code duplication.

aalexandrov commented on emma-spark/src/main/scala/org/emmalanguage/api/SparkRDD.scala:
```diff
@@ -240,4 +240,15 @@ object SparkRDD {
+  )(implicit spark: SparkSession): DataBag[(K, B)] = xs match {
+    case xs: SparkRDD[A] => xs.rep
+      .map(x => key(x) -> sng(x)).reduceByKey(uni)
+    case xs: SparkDataset[A] => xs.rep
```
Call `SparkDataset.foldGroup` in order to avoid code duplication.

aalexandrov commented on emma-spark/src/main/scala/org/emmalanguage/api/SparkDataset.scala:
```diff
@@ -214,4 +214,15 @@ object SparkDataset {
+      .map(x => key(x) -> sng(x)).groupByKey(_._1)
+      .reduceGroups((x, y) => x._1 -> uni(x._2, y._2))
+      .map((kv: (K, (K, B))) => kv._2)
+  }
```
Implement this case by going through an `rdd` conversion for the moment (this will avoid group materialization as opposed to the `reduceGroups` variant).

```scala
case xs: SparkDataset[A] => xs.rep.rdd
  .map(x => key(x) -> sng(x)).reduceByKey(uni)).toDS
```

joroKr21 commented on emma-language/src/main/scala/org/emmalanguage/compiler/lang/opt/FoldGroupFusion.scala:
```diff
@@ -0,0 +1,409 @@
+        // Iterate to combine trees in the forest while possible.
+        // We can't forget about trees that don't change in a single step,
+        // because a squash might enable a fusion and vice versa.
+        Iterator.iterate(forest, true) { case (forest, _) =>
```
Unfortunately using a variable as the convergence criteria doesn't fit together with using a `Stream` to hold the forest. I would have to convert back and forth between a strict collection (such as a `Vector` or `List` on every step).

aalexandrov replied:
I don't get where the problem is - `forest` remains a `Stream` in the suggested rewrite.

joroKr21 replied:
The condition `forestDidGrow` sometimes has the wrong value due to laziness of the stream. It's updated inside the `flatMap` call, which is lazy.

aalexandrov replied:
OK, the `flatMap` is on a Stream because of the preceeding `toStream` call, maybe just do a `flatMap` on the result of the `groupBy` (a `Map`) and convert to `Stream` at the end.

joroKr21 commented:
Ok, I addressed the comments:
 - Nomenclature - renamed as suggested
 - Readability:
   - Converted the fixpoint iteration to a while loop
   - Split up the generation of `emp`, `sng` and `uni` in the banana-fusion case
   - Added comments that show the transformations as source code
- In addition: extra 3 LOC to handle > 22 folds via nested tuples.
